main/inc/lib/myspace.lib.php

```diff
@@ -4020,86 +4407,80 @@ protected static function getCompanyLearnpathSubscription(
+            ON ( u.id = ip.to_user_id )
+            INNER JOIN $tblLp AS lp_table
+            ON ( lp_table.iid = ip.ref )
+            INNER JOIN $tblLpItem AS lp_table_item
+            ON ( lp_table.id = lp_table_item.lp_id )
             WHERE
-                lastedit_type = 'LearnpathSubscription'
-
-                ";
+                ip.lastedit_type = 'LearnpathSubscription' ";
             if (strlen($whereCondition) > 2) {
                 $query .= $whereCondition;
             }
-            $queryResult = Database::query($query);
-            while ($row = Database::fetch_array($queryResult, 'ASSOC')) {
-                // $courseId = (int)$row['c_id'];
-                $studentId = (int) $row['to_user_id'];
-                $company = isset($row['company']) ? $row['company'] : '';
-                if ($company == '') {
-                    $company = get_lang('NoEntity');
-                }
-                // $lpId = $row['ref'];
-                if ($lpId != 0 && $studentId != 0) {
-                    if ($whitCompany == true) {
-                        $companys[] = [
-                            'id' => $studentId,
-                            'company' => $company,
-                        ];
-                    } else {
-                        $companys[] = $studentId;
-                    }
-                } else {
-                    $companys[$company][] = $studentId;
-                    $companys[$company] = array_unique($companys[$company]);
+            if ($withGroups) {
+                $query = "
+                SELECT DISTINCT
+                    ip.ref AS lp_item,
+                    lp_table_item.iid AS lp_item_id,
+                    ip.session_id AS session_id,
+                    ip.lastedit_type AS type,
+                    u.username AS username,
+                    ip.lastedit_date AS lastedit_date,
+                    ip.to_group_id AS group_id,
+                    ug.user_id AS id
+                FROM
+                    $tblItemProperty AS ip
+                INNER JOIN $tblGroupUser AS ug
+                ON ( ug.group_id = ip.to_group_id )
+                INNER JOIN `$tblUser` AS u
+                ON ( u.id = ug.user_id )
+                INNER JOIN $tblLp AS lp_table
+                ON ( lp_table.iid = ip.ref )
+                INNER JOIN $tblLpItem AS lp_table_item
+                ON ( lp_table.id = lp_table_item.lp_id )
+                WHERE
+                    ip.lastedit_type = 'LearnpathSubscription' AND
+                    ip.to_group_id != 0
+                ";
+                if (strlen($whereCondition) > 2) {
+                    $query .= $whereCondition;
                 }
             }
+            $query .= ' ORDER BY ip.ref, ip.session_id ';
+            $queryResult = Database::query($query);
+            $data = Database::store_result($queryResult, 'ASSOC');
+            $totalData = count($data);
+            for ($i = 0; $i < $totalData; $i++) {
```

Comment:
foreach

Reply:
Se usa for a propósito para mejorar el rendimiento ya que este reporte, en un momento, fue extremadamente lento, puedes ver la comparación de for vs foreach aqui https://www.phpbench.com/  en la sección "Modify Loop: foreach() vs. for vs. while(list() = each())"

Reply:
Cuál fue la mejora para este caso?

Reply:
Mejoras en sql primero, aun con bajo rendimiento, cambio de foreach a for y mejoro el rendimiento.

Reply:
Cuál fue la mejora del cambio de foreach a for? Cuantos segundos/minutos se ganó?

Reply:
de 10 min en local bajo a 3 con todos los elementos aplicados